src/pudl/transform/classes.py

```diff
@@ -0,0 +1,504 @@
+        """Require that all unit conversions result in no column renaming.
+
+        This constraint is imposed so that the same unit conversion definitions
+        can be re-used both for unit corrections and columnwise unit conversions.
+        """
+        new_conversions = []
+        for conv in v:
+            new_conversions.append(
+                UnitConversion(multiplier=conv.multiplier, adder=conv.adder)
+            )
+        return new_conversions
+
+
+class TableTransformParams(TransformParams):
+    """All the generic transformation parameters for a table.
+
+    Data source specific TableTransformParams can be defined by models that inherit
+    from this one in the data source specific transform modules.
+
+    """
+
+    rename_columns: RenameColumns = {}
+    convert_units: dict[str, UnitConversion] = {}
+    categorize_strings: dict[str, StringCategories] = {}
+    nullify_outliers: dict[str, ValidRange] = {}
+    normalize_strings: dict[str, bool] = {}
+    correct_units: list[UnitCorrections] = []
+
+    @classmethod
+    def from_id(cls, table_id: enum.Enum) -> "TableTransformParams":
+        """A factory method that looks up transform parameters based on table_id."""
+        return cls(**pudl.transform.TRANSFORM_PARAMS[table_id.value])
+
+
+#####################################################################################
+# Transform Protocol & General Function Definitions
+#####################################################################################
+class ColumnTransformFn(Protocol):
+    """Callback protocol defining a per-column transformation function."""
+
+    def __call__(self, col: pd.Series, params: TransformParams) -> pd.Series:
+        """Create a callable."""
+        ...
+
+
+class TableTransformFn(Protocol):
+    """Callback protocol defining a per-table transformation function."""
+
+    def __call__(self, df: pd.DataFrame, params: TransformParams) -> pd.DataFrame:
+        """Create a callable."""
+        ...
+
+
+class MultiColumnTransformFn(Protocol):
+    """Callback protocol defining a per-table transformation function."""
+
+    def __call__(
+        self, df: pd.DataFrame, params: MultiColumnTransformParams
+    ) -> pd.DataFrame:
+        """Create a callable."""
+        ...
+
+
+def multicol_transform_fn_factory(
+    col_fn: ColumnTransformFn,
+    drop=True,
+) -> MultiColumnTransformFn:
+    """A factory for creating a multi-column transform function."""
+
+    class InnerMultiColumnTransformFn(
+        Callable[[pd.DataFrame, MultiColumnTransformParams], pd.DataFrame]
+    ):
+        __name__ = col_fn.__name__ + "_multicol"
+
+        def __call__(
+            self, df: pd.DataFrame, params: MultiColumnTransformParams
+        ) -> pd.DataFrame:
+            drop_col: bool = drop
+            for col_name in params:
+                if col_name in df.columns:
+                    logger.debug(f"Applying {col_fn.__name__} to {col_name}")
```

Comment:
I like this but can we keep the format of the loggers semi consistent and put `{table name}:` first?

Reply:
In this case we can't, because this function is not defined within the context of the table transformer classes, so it has no idea what table it's applying to. This is one reason I ultimately wanted to wrap all of the multi-column transform functions in methods within the abstract base class -- so that they can automatically say "Hey I'm gonna do this table now." before all of these column-level logging messages are generated.